src/WebJobs.Extensions.DurableTask/Listener/OutOfProcOrchestrationShim.cs

```diff
@@ -104,7 +99,135 @@ internal async Task<bool> ScheduleDurableTaskEvents(OrchestrationInvocationResul
+                        task = fireAndForgetTask;
+                        break;
+                    }
+
+                case AsyncActionType.ContinueAsNew:
+                    this.context.ContinueAsNew(action.Input);
+                    task = fireAndForgetTask;
+                    break;
+                case AsyncActionType.WaitForExternalEvent:
+                    task = this.context.WaitForExternalEvent<object>(action.ExternalEventName);
+                    break;
+                case AsyncActionType.CallHttp:
+                    task = this.context.CallHttpAsync(action.HttpRequest);
+                    break;
+                case AsyncActionType.WhenAll:
+                    task = Task.WhenAll(action.CompoundActions.Select(x => this.InvokeAPIFromAction(x)));
+                    break;
+                case AsyncActionType.WhenAny:
+                    task = Task.WhenAny(action.CompoundActions.Select(x => this.InvokeAPIFromAction(x)));
+                    break;
+                default:
+                    throw new Exception($"Received an unexpected action type from the out-of-proc function: '${action.ActionType}'.");
+            }
+
+            return task;
+        }
+
+        /// <summary>
+        /// Replays the orchestration execution from an OOProc SDK in .NET.
+        /// </summary>
+        /// <param name="actions">The OOProc actions payload.</param>
+        /// <returns>An awaitable Task that completes once replay completes.</returns>
+        private async Task ProcessAsyncActionsV2(AsyncAction[] actions)
+        {
+            foreach (AsyncAction action in actions)
+            {
+                await this.InvokeAPIFromAction(action);
+            }
+        }
+
+        /// <summary>
+        /// Replays the OOProc orchestration based on the actions array. It uses the schema enum to
+        /// determine which replay implementation is most appropiate.
+        /// </summary>
+        /// <param name="actions">The OOProc actions payload.</param>
+        /// <param name="schema">The OOProc protocol schema version.</param>
+        /// <returns>An awaitable Task that completes once replay completes.</returns>
+        private async Task ReplayOOProcOrchestration(AsyncAction[][] actions, SchemaVersion schema)
+        {
+            switch (schema)
+            {
+                case SchemaVersion.V2:
+                    // In this schema, action arrays should be 1 dimensional (1 action per yield), but due to legacy behavior they're nested within a 2-dimensional array.
+                    if (actions.Length != 1)
+                    {
+                        throw new ArgumentException($"With OOProc schema {schema}, expected actions array to be of length 1 in outer layer but got size: {actions.Length}");
+                    }
+
+                    await this.ProcessAsyncActionsV2(actions[0]);
+                    break;
+                default: // same as: case SchemaVersion.Original
```

Comment:
I wonder if you should treat `default` as an error case and have `SchemaVersion.Original` be handled explicitly. That way if we release a V3 schema and someone tries to run it on an old extension, we can correctly catch the problem and throw a meaningful exception.

Reply:
Good call, done!